controllers/nstemplateset/cluster_resources.go

```diff
@@ -247,6 +247,12 @@ func (r *clusterResourcesManager) deleteClusterResource(ctx context.Context, nsT
 	if err := r.setStatusUpdatingIfNotProvisioning(ctx, nsTmplSet); err != nil {
 		return false, err
 	}
+	//remove finalizer before deleting
+	toDelete.SetFinalizers(nil)
```

Comment:
Why?

Reply:
because the object will not be deleted untill there is a finalizer. fakeClient.Delete now skips the deletion if a finalizer is present.

Reply:
Hm... but in the runtime, isn't the expected workflow is the following:
- If there is no finalizer then add one (if the resource is not already being deleted).
- If there is a finalizer present and the resource is being deleted (there is a deletion timestamp) then do all required cleanup and then delete the finalizer to unblock the actual resource deletion by the garbage collector. Otherwise the resource will be stuck in the "terminating" if a finalizesr is present.

Or I'm missing something?

Reply:
As i understand it:
- it doesn't add a finalizer if not present, rather it errors out if a finalizer is not present when trying to create a resource with deletion timestamp.
- When a finalizer is present and Delete() is called - it adds a deletion timestamp and updates as [here](https://github.com/lleshchi/controller-runtime/blob/7a66d580c0c53504f5b509b45e9300cc18a1cc30/pkg/client/fake/client.go#L1068-L1073). And update actually deletes the resource only when there is no finalizer as [here](https://github.com/lleshchi/controller-runtime/blob/7a66d580c0c53504f5b509b45e9300cc18a1cc30/pkg/client/fake/client.go#L454-L456). Thus the action of removing the finalizer is to be done in our test. 
When i draw a parallel to the real client as well - we remove the finalizers in our controllers after all the required cleanup.

Reply:
@ranakan19 but what you are changing here is the production code, not a test. We shouldn't remove any finalizers we are not responsible for - then only finalizer we can add/remove (in production code) is the toolchain one. Nothing else.
If you need to do some workarounds around finalizers to make FakeClient working, then do it in test files, not in the production code.

Reply:
I mislooked that it not in the test, i'll make the necessary changes

Reply:
@alexeykazakov @MatousJobanek removed in [b1816ef](https://github.com/codeready-toolchain/member-operator/pull/551/commits/b1816ef8297f6b49fba73cf142878768e514bae1)